src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -330,37 +330,109 @@ public static unsafe SocketError Receive(SafeCloseSocket handle, byte[] buffer,
+            {
+                address = new IPAddress((long)controlBuffer->address);
+            }
 
-            SocketError errorCode = SocketError.Success;
+            return new IPPacketInformation(address, (int)controlBuffer->index);
+        }
+
+        public static unsafe IPPacketInformation GetIPPacketInformation(Interop.Winsock.ControlDataIPv6* controlBuffer)
+        {
+            IPAddress address = IPAddress.IPv6None;
+            if (controlBuffer->length != UIntPtr.Zero)
+            {
+                var addressArray = new byte[Interop.Winsock.IPv6AddressLength];
+                Marshal.Copy((IntPtr)(controlBuffer->address), addressArray, 0, Interop.Winsock.IPv6AddressLength);
+                address = new IPAddress(addressArray);
+            }
+
+            return new IPPacketInformation(address, (int)controlBuffer->index);
+        }
+
+        public static unsafe SocketError ReceiveMessageFrom(Socket socket, SafeCloseSocket handle, byte[] buffer, int offset, int size, ref SocketFlags socketFlags, Internals.SocketAddress socketAddress, out Internals.SocketAddress receiveAddress, out IPPacketInformation ipPacketInformation, out int bytesTransferred)
+        {
+            bool ipv4, ipv6;
+            Socket.GetIPProtocolInformation(socket.AddressFamily, socketAddress, out ipv4, out ipv6);
 
             bytesTransferred = 0;
-            try
+            receiveAddress = socketAddress;
+            ipPacketInformation = default(IPPacketInformation);
+
+            fixed (byte * ptrBuffer = buffer)
+            fixed (byte * ptrSocketAddress = socketAddress.Buffer)
             {
-                // This can throw ObjectDisposedException (retrieving the delegate AND resolving the handle).
-                if (socket.WSARecvMsgBlocking(
-                    handle.DangerousGetHandle(),
-                    Marshal.UnsafeAddrOfPinnedArrayElement(asyncResult._messageBuffer, 0),
-                    out bytesTransferred,
-                    IntPtr.Zero,
-                    IntPtr.Zero) == SocketError.SocketError)
+                Interop.Winsock.WSAMsg wsaMsg;
+                wsaMsg.socketAddress = (IntPtr)ptrSocketAddress;
+                wsaMsg.addressLength = (uint)socketAddress.Size;
+                wsaMsg.flags = socketFlags;
+
+                WSABuffer wsaBuffer;
+                wsaBuffer.Length = size;
+                wsaBuffer.Pointer = (IntPtr)(ptrBuffer + offset);
+                wsaMsg.buffers = (IntPtr)(&wsaBuffer);
+                wsaMsg.count = 1;
+
+                if (ipv4)
                 {
-                    errorCode = GetLastSocketError();
+                    Interop.Winsock.ControlData controlBuffer;
+                    wsaMsg.controlBuffer.Pointer = (IntPtr)(&controlBuffer);
+                    wsaMsg.controlBuffer.Length = sizeof(Interop.Winsock.ControlData);
+
+                    if (socket.WSARecvMsgBlocking(
+                        handle.DangerousGetHandle(),
```

Comment:
It just occurred to me that the previous DangerousGetHandle approach may have been by-design.  Not addref/releasing for a synchronous call would allow the handle to be closed while it was in use, and it's possible the intention was to allow a blocking receive call to be unblocked by closing the corresponding socket.  If that is the case and that's being relied on, changing it to just pass the SafeHandle would break that, as the close would be delayed until after the receive completed.  Then again, not using a safe handle in such a case also opens the system up to handle recycling problems.  @cipop, do you know what's intended here?